api/routes/websoc/schema.ts

```diff
@@ -0,0 +1,127 @@
+import {
+  anyArray,
+  cancelledCoursesOptions,
+  divisionCodes,
+  fullCoursesOptions,
+  geCodes,
+  quarters,
+  sectionTypes,
+} from "peterportal-api-next-types";
+import { z } from "zod";
+
+/**
+ * Input to a transform function.
+ */
+type TransformInput = string | string[] | undefined;
+
+/**
+ * Get unique, sorted array of strings.
+ * @param value String of comma-separated values or array of such strings.
+ */
+function flattenStringsAndSplit(value: TransformInput): string[] | undefined {
+  if (!value) return undefined;
+  const unique = new Set(
+    Array.isArray(value) ? value.flatMap((x) => x.split(",")) : value.split(",")
+  );
+  return [...unique].sort();
+}
+
+const days = ["Su", "M", "Tu", "W", "Th", "F", "Sa"];
+
+/**
+ * Get unique, sorted array of day strings from input.
+ * @param value String of combined days of the week (e.g. ``MWF``) or array of such strings.
+ */
+function flattenDayStringsAndSplit(
+  value: TransformInput
+): string[] | undefined {
+  if (!value) return undefined;
+  const unique = new Set(
+    Array.isArray(value)
+      ? value.flatMap((x) => days.filter((y) => y.includes(x)))
+      : days.filter((x) => value.includes(x))
+  );
+  return [...unique].sort();
+}
+
+/**
+ * Parse an unknown query to the websoc endpoint.
+ */
+export const QuerySchema = z
+  .object({
+    year: z
+      .string({ required_error: 'Parameter "year" not provided' })
+      .length(4, { message: "Invalid year provided" }),
+    quarter: z.enum(quarters, {
+      required_error: 'Parameter "quarter" not provided',
+      invalid_type_error: "Invalid quarter provided",
+    }),
+    ge: z.enum(anyArray).or(z.enum(geCodes)).optional(),
+    department: z.string().optional(),
+    courseTitle: z.string().optional(),
+    courseNumber: z
+      .string()
+      .array()
+      .or(z.string())
+      .optional()
+      .transform(flattenStringsAndSplit),
+    sectionCodes: z
+      .string()
+      .array()
+      .or(z.string())
+      .optional()
+      .transform(flattenStringsAndSplit)
+      .optional(),
```

Comment:
```suggestion
      .transform(flattenStringsAndSplit),
```
`flattenStringsAndSplit` says it can return undefined, so I think this is redundant. Same for any other `optional` that are chained after `transform`.